src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -357,149 +278,265 @@ export async function getV3CandidatePools({
-          subgraphPool.token1.id == tokenInAddress)
-      );
-    })
-    .slice(0, topNTokenInOut)
-    .value();
+    if (
+      topByBaseWithTokenOut.length < topNWithBaseToken &&
+      (
+        (baseTokensAddresses.includes(subgraphPool.token0.id) && subgraphPool.token1.id == tokenOutAddress) ||
+        (baseTokensAddresses.includes(subgraphPool.token1.id) && subgraphPool.token0.id == tokenOutAddress)
+      )
+    ) {
+      topByBaseWithTokenOut.push(subgraphPool);
+      continue;
+    }
 
-  addToAddressSet(topByTVLUsingTokenIn);
+    // Main reason we need this is for gas estimates, only needed if token out is not native.
+    // We don't check the seen address set because if we've already added pools for getting native quotes
+    // there's no need to add more.
+    if (
+      topByEthQuoteTokenPool.length < 2 &&
+      (
+        (
+          WRAPPED_NATIVE_CURRENCY[chainId]?.symbol == WRAPPED_NATIVE_CURRENCY[ChainId.MAINNET]?.symbol &&
+          tokenOut.symbol != 'WETH' &&
+          tokenOut.symbol != 'WETH9' &&
+          tokenOut.symbol != 'ETH'
+        ) ||
+        (
+          WRAPPED_NATIVE_CURRENCY[chainId]?.symbol == WMATIC_POLYGON.symbol &&
+          tokenOut.symbol != 'MATIC' &&
+          tokenOut.symbol != 'WMATIC'
+        )
+      ) &&
+      (
+        routeType === TradeType.EXACT_INPUT && (
+          (subgraphPool.token0.id == wrappedNativeAddress && subgraphPool.token1.id == tokenOutAddress) ||
+          (subgraphPool.token1.id == wrappedNativeAddress && subgraphPool.token0.id == tokenOutAddress)
+        ) ||
+        routeType === TradeType.EXACT_OUTPUT && (
+          (subgraphPool.token0.id == wrappedNativeAddress && subgraphPool.token1.id == tokenInAddress) ||
+          (subgraphPool.token1.id == wrappedNativeAddress && subgraphPool.token0.id == tokenInAddress)
+        )
+      )
+    ) {
+      topByEthQuoteTokenPool.push(subgraphPool);
+      continue;
+    }
 
-  const topByTVLUsingTokenOut = _(subgraphPoolsSorted)
-    .filter((subgraphPool) => {
-      return (
-        !poolAddressesSoFar.has(subgraphPool.id) &&
-        (subgraphPool.token0.id == tokenOutAddress ||
-          subgraphPool.token1.id == tokenOutAddress)
-      );
-    })
-    .slice(0, topNTokenInOut)
-    .value();
+    if (topByTVL.length < topN) {
+      topByTVL.push(subgraphPool);
+      continue;
+    }
+
+    if (
+      topByTVLUsingTokenIn.length < topNTokenInOut &&
+      (subgraphPool.token0.id == tokenInAddress || subgraphPool.token1.id == tokenInAddress)
+    ) {
+      topByTVLUsingTokenIn.push(subgraphPool);
+      continue;
+    }
 
+    if (
+      topByTVLUsingTokenOut.length < topNTokenInOut &&
+      (subgraphPool.token0.id == tokenOutAddress || subgraphPool.token1.id == tokenOutAddress)
+    ) {
+      topByTVLUsingTokenOut.push(subgraphPool);
+      continue;
+    }
+  }
+
+  // Add the addresses found so far to our address set
+  addToAddressSet(topByBaseWithTokenIn);
```

Comment:
I think we might want to check `poolAddressesSoFar` dedup in `addToAddressSet` with current implementation.

Reply:
Also helps on perf, since we can dedup across `topByBaseWithTokenIn`, ``topByBaseWithTokenOut`... `topByTVL `

Reply:
nvm, with the current for-loop each if-continue structure, we are guaranteed to have no duplicates.